snuba/datasets/dataset.py

```diff
@@ -56,6 +62,9 @@ def get_all_entities(self) -> Sequence[Entity]:
     def get_query_pipeline_builder(self) -> DatasetQueryPipelineBuilder:
         return DatasetQueryPipelineBuilder()
 
+    def get_quota_control_policy(self) -> QuotaControlPolicy:
+        return self.__quota_control
+
```

Comment:
While I think the Dataset is the correct place for this, I'm a little confused at your reasoning. You have to parse the query into the AST in order to find the project IDs, so this is not preventing any of that. I think it makes sense having this check at the Dataset level since it would be weird to have a JOIN where one entity was rate limited but the other wasn't.

Reply:
> You have to parse the query into the AST in order to find the project IDs, so this is not preventing any of that.

The plan for step three was to make the "quota identifier" which is basically an abstract key to identify the quota to apply (project id in this case) to be passed by the client outside the SnQL query in order to only have to parse the JSON. 
Ideally we should not even parse the json and have that as a header so we can reject even earlier. 

So this is how I was planning to skip parsing entirely.

Reply:
Got it, that makes more sense. We should build that into the SDK so it's not something the product teams have to think about.